ALEX P Week2 Databases

op117 opened the pull request:
@dyaboykyl Hi,
updated the solution to the tasks of the second week taking into account the comments in the tasks of the first week of the module:
- Removed commits of preparatory exercises;
- async/await with promises.

dyaboykyl commented on Week2/ex4-Aggregate-Functions.js:
```diff
@@ -0,0 +1,52 @@
+    console.table(authorsPerPaperResults);
+
+    const papersByFemaleAuthors = `
+      SELECT COUNT(author_paper.paper_id) AS total_papers
+      FROM authors
+      INNER JOIN author_paper ON authors.author_id = author_paper.author_id
+      WHERE authors.gender = 'female';
```
What if a paper had more than one female author? Would any problem arise with this query?

dyaboykyl commented on Week2/ex4-Aggregate-Functions.js:
```diff
@@ -0,0 +1,52 @@
+    const [avgHIndexResults] = await connection.query(avgHIndex);
+    console.log("Average h-index per university:");
+    console.table(avgHIndexResults);
+
```
Looks like you're missing queries for questions 3 and 4 in this exercise. Please double check the assignment

dyaboykyl reviewed (changes requested):
Overall the code looks good! Just a couple of comments

op117 commented:
Hi @dyaboykyl !
Thanks for your feedback. I have updated my solution based on your comments:
- missing queries added;
- moved duplicate code to executeQuery function;
- closing the connection in case of error;
- using Top-Level Await (Node v20+);

dyaboykyl commented on Week2/ex4-Aggregate-Functions.js:
```diff
@@ -1,52 +1,77 @@
+    await executeQuery(totalPapersPerUniversity, "Total papers per university:");
+
+    // Query 5
+    const minMaxHIndex = `
+        SELECT university, MIN(h_index) AS min_h_index, MAX(h_index) AS max_h_index
+        FROM authors
+        GROUP BY university;
```
Would any problems arise if one paper had multiple authors from the same university?

dyaboykyl replied:
Sorry Alex my fault, I put this comment on the wrong query. In fact with this query there is no possibility of duplicating h indexes because you are not joining with the author_paper relationship table. So this query did not need to change.

The query that has a possibility of duplicate values is the Query 4, total papers per university. Since that query is using a join on a the author_paper table, any paper that has multiple authors from the same university will counted more than once, thus it needs a DISTINCT operator on the university

Again, my bad for putting the comment in the wrong spot.

dyaboykyl commented on Week2/ex1-Keys.js:
```diff
@@ -0,0 +1,43 @@
+require('dotenv').config();
+const mysql = require('mysql2/promise');
+
+(async () => {
+  try {
```
The top level anonymous functions can be removed in all of the files

op117 commented:
Hi @dyaboykyl.
Fixed.

dyaboykyl commented:
I updated my previous comment, please take a look - it should be a small fix. Otherwise the code looks great!

op117 commented:
Hi @dyaboykyl.
Updated.

dyaboykyl reviewed (approved):
Looks good!
